cmd/nodecmd/whitelist.go

```diff
@@ -0,0 +1,170 @@
+	for _, cloudSecurityGroup := range cloudSecurityGroupList {
+		if cloudSecurityGroup.cloud == constants.GCPCloudService {
+			if !slices.Contains(gcpSGList, cloudSecurityGroup.securityGroup) {
+				gcpSGList = append(gcpSGList, cloudSecurityGroup.securityGroup)
+			}
+			continue
+		}
+		ux.Logger.PrintToUser("Whitelisting IP %s in %s cloud region %s", userIPAddress, cloudSecurityGroup.cloud, cloudSecurityGroup.region)
+		if cloudSecurityGroup.cloud == "" || cloudSecurityGroup.cloud == constants.AWSCloudService {
+			ec2Svc, err := awsAPI.NewAwsCloud(awsProfile, cloudSecurityGroup.region)
+			if err != nil {
+				ux.Logger.PrintToUser("Failed to establish connection to %s cloud region %s", cloudSecurityGroup.cloud, cloudSecurityGroup.region)
+				return err
+			}
+			securityGroupExists, sg, err := ec2Svc.CheckSecurityGroupExists(cloudSecurityGroup.securityGroup)
+			if err != nil || !securityGroupExists {
+				ux.Logger.PrintToUser("Can't find security group %s in %s cloud region %s", cloudSecurityGroup.securityGroup, cloudSecurityGroup.cloud, cloudSecurityGroup.region)
+				return err
+			}
+			ipInTCP := awsAPI.CheckUserIPInSg(&sg, userIPAddress, constants.SSHTCPPort)
+			ipInHTTP := awsAPI.CheckUserIPInSg(&sg, userIPAddress, constants.AvalanchegoAPIPort)
+			var alreadyWhitelisted bool
+			if ipInTCP {
+				alreadyWhitelisted = true
+			} else {
+				if err := ec2Svc.AddSecurityGroupRule(*sg.GroupId, "ingress", "tcp", userIPAddress, constants.SSHTCPPort); err != nil {
+					ux.Logger.PrintToUser("Failed to whitelist IP %s in %s cloud region %s for ssh access", userIPAddress, cloudSecurityGroup.cloud, cloudSecurityGroup.region)
+					return err
+				}
+			}
+			if ipInHTTP {
+				alreadyWhitelisted = true
+			} else {
+				if err := ec2Svc.AddSecurityGroupRule(*sg.GroupId, "ingress", "tcp", userIPAddress, constants.AvalanchegoAPIPort); err != nil {
+					ux.Logger.PrintToUser("Failed to whitelist IP %s in %s cloud region %s for http access", userIPAddress, cloudSecurityGroup.cloud, cloudSecurityGroup.region)
+					return err
+				}
+			}
+			if alreadyWhitelisted {
+				ux.Logger.PrintToUser("IP %s is already whitelisted in %s cloud region %s. Skipping...", userIPAddress, cloudSecurityGroup.cloud, cloudSecurityGroup.region)
+			}
+		}
+	}
+	if len(gcpSGList) > 0 {
+		networkName := fmt.Sprintf("%s-network", defaultAvalancheCLIPrefix())
+		gcpClient, projectName, _, err := getGCPCloudCredentials()
+		if err != nil {
+			return err
+		}
+		gcpCloud, err := gcpAPI.NewGcpCloud(gcpClient, projectName, context.Background())
+		if err != nil {
+			return err
+		}
+		ux.Logger.PrintToUser("Whitelisting IP %s in %s cloud", userIPAddress, constants.GCPCloudService)
+		_, err = gcpCloud.SetFirewallRule(userIPAddress, fmt.Sprintf("%s-%s", networkName, strings.ReplaceAll(userIPAddress, ".", "")), networkName, []string{strconv.Itoa(constants.SSHTCPPort), strconv.Itoa(constants.AvalanchegoAPIPort)})
+		if err != nil && errors.IsAlreadyExists(err) {
```

Comment:
probably add a second if inside and keep one external if for err != nil

Reply:
good one. I like it. fixed

Reply:
agree